File: mobile/modules/gizu-signer/android/src/main/java/io/gizu/signer/GizuSignerModule.kt
package io.gizu.signer

import android.app.AlertDialog
import android.util.Base64
import androidx.credentials.*
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import kotlinx.coroutines.*
import org.json.JSONArray
import org.json.JSONObject
import java.security.SecureRandom
import uniffi.gizu_signer_core.*

/** Fixed-message development probe; never accepts PRF, messages or payloads from JS. */
class GizuSignerModule : Module() {
  private var busy = false
  private var task: Job? = null
  private var dialog: AlertDialog? = null
  private var cancelPending: (() -> Unit)? = null
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
  private fun random() = ByteArray(32).also { SecureRandom().nextBytes(it) }
  private fun b64(value: ByteArray) = Base64.encodeToString(value, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
  private val salt = "896d46ac4ac191885c46137439db7bb52fb05cff3ecd34af7cdae0a1e0c00db9"
  override fun definition() = ModuleDefinition {
    Name("GizuSigner")
    Function("getCapabilities") {
      val debug = ((appContext.reactContext?.applicationInfo?.flags ?: 0) and android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE) != 0
      mapOf("nativeTransfers" to (debug && android.os.Build.VERSION.SDK_INT >= 28), "fixedProbe" to (debug && android.os.Build.VERSION.SDK_INT >= 28))
    }
    AsyncFunction("executeOperation") { proposal: String, promise: Promise ->
      scope.launch {
        val activity = appContext.currentActivity
        if (busy || activity == null || android.os.Build.VERSION.SDK_INT < 28 ||
          (activity.applicationInfo.flags and android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE) == 0) {
          promise.reject("UNAVAILABLE", "Native testnet operation unavailable", null); return@launch
        }
        try {
          busy = true
          TransferHost.start(activity, proposal) { result ->
            busy = false
            if (result == null) promise.reject("OPERATION_STOPPED", "Operation stopped. Refresh native status before retrying; submitted transfers cannot be undone.", null)
            else promise.resolve(result)
          }
        } catch (_: Exception) { busy = false; promise.reject("INVALID_OPERATION", "Native operation rejected", null) }
      }
    }
    AsyncFunction("getOperationStatus") { promise: Promise ->
      scope.launch {
        if (busy) { promise.reject("BUSY", "Native operation in progress", null); return@launch }
        val context = appContext.reactContext
        if (context == null) { promise.reject("UNAVAILABLE", "Native status unavailable", null); return@launch }
        busy = true
        try { promise.resolve(TransferJournal(context).reconcile(MonadRpc())) }
        catch (_: Exception) { promise.reject("RPC_UNAVAILABLE", "Cannot reconcile native status. Do not repeat uncertain transfers.", null) }
        finally { busy = false }
      }
    }
    Function("cancelOperation") { scope.launch { TransferHost.cancel() } }
    Function("lock") { scope.launch { TransferHost.cancel(); cancelPending?.invoke() } }
    AsyncFunction("openNativeProbe") { promise: Promise ->
      scope.launch {
        val activity = appContext.currentActivity
        if (busy || android.os.Build.VERSION.SDK_INT < 28 || activity == null || (activity.applicationInfo.flags and android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE) == 0) {
          promise.reject("UNAVAILABLE", "Native development probe unavailable", null)
          return@launch
        }
        busy = true
        var finished = false
        var expiry: Job? = null
        fun finishError() { if (finished) return; finished = true; expiry?.cancel(); busy = false; cancelPending = null; dialog?.dismiss(); dialog = null; promise.reject("PROBE_FAILED", "Native probe cancelled, unsupported or failed. Open an existing credential after partial creation.", null) }
        cancelPending = { task?.cancel(); finishError() }
        expiry = scope.launch { delay(120_000); cancelPending?.invoke() }
        fun start(create: Boolean) {
          if (finished) return
          task = scope.launch {
            try {
              val proofs = withTimeout(120_000) {
                val manager = CredentialManager.create(activity)
                var createdId: String? = null
                if (create) {
                  val request = JSONObject().put("challenge", b64(random()))
                    .put("rp", JSONObject().put("id", "gizu.io").put("name", "Gizu"))
                    .put("user", JSONObject().put("id", b64(random())).put("name", "gizu-native-test").put("displayName", "Gizu native test"))
                    .put("pubKeyCredParams", JSONArray().put(JSONObject().put("type", "public-key").put("alg", -7)))
                    .put("authenticatorSelection", JSONObject().put("residentKey", "required").put("userVerification", "required"))
                    .put("attestation", "none")
                    .put("extensions", JSONObject().put("prf", JSONObject()))
                  // Registration may leave a credential on cancellation of the subsequent assertion.
                  val registration = manager.createCredential(activity, CreatePublicKeyCredentialRequest(request.toString())) as CreatePublicKeyCredentialResponse
                  createdId = JSONObject(registration.registrationResponseJson).getString("id")
                }
                val saltBytes = salt.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
                val request = JSONObject().put("challenge", b64(random())).put("rpId", "gizu.io")
                  .put("userVerification", "required")
                  .put("extensions", JSONObject().put("prf", JSONObject().put("eval", JSONObject().put("first", b64(saltBytes)))))
                createdId?.let { request.put("allowCredentials", JSONArray().put(JSONObject().put("type", "public-key").put("id", it))) }
                val response = manager.getCredential(activity, GetCredentialRequest(listOf(GetPublicKeyCredentialOption(request.toString()))))
                val credential = response.credential as? PublicKeyCredential ?: error("unsupported")
                // Provider JSON remains native. JVM strings cannot be reliably erased.
                val result = JSONObject(credential.authenticationResponseJson)
                val prf = Base64.decode(result.getJSONObject("clientExtensionResults").getJSONObject("prf").getJSONObject("results").getString("first"), Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
                try {
                  ensureActive()
                  if (activity.isFinishing || activity.isDestroyed || finished) error("not foreground")
                  // Only this private native path accepts secret bytes. No bridge export.
                  runNativeProbe(prf, random().joinToString("") { "%02x".format(it) })
                } finally { prf.fill(0) }
              }
              val publicResults = proofs.map { mapOf("accountIndex" to it.accountIndex.toInt(), "address" to it.address, "message" to it.message, "signature" to it.signatureHex) }
              if (finished) return@launch
              finished = true; expiry?.cancel(); busy = false; cancelPending = null; dialog = null
              promise.resolve(publicResults)
            } catch (_: Exception) { finishError() }
          }
        }
        dialog = AlertDialog.Builder(activity)
          .setTitle("Native compatibility test")
          .setMessage("Use an unfunded test passkey. Approve 16 fixed test signatures across accounts 0–15. No transactions, login or spending. Keys stay in native memory and are released after this test. Creation may require two prompts.")
          .setPositiveButton("Open existing") { _, _ -> start(false) }
          .setNeutralButton("Create test passkey") { _, _ -> start(true) }
          .setNegativeButton("Cancel") { _, _ -> finishError() }
          .setOnCancelListener { finishError() }.show()
      }
    }
    Function("cancelProbe") { scope.launch { cancelPending?.invoke() } }
    OnDestroy { scope.launch { TransferHost.cancel(); cancelPending?.invoke(); scope.cancel() } }
  }
}

File: mobile/modules/gizu-signer/ios/GizuSignerModule.swift
import ExpoModulesCore
import AuthenticationServices
import UIKit
import CryptoKit

public final class GizuSignerModule: Module {
  private var probe: AnyObject?
  private var transfer: AnyObject?
  private var reconciling = false
  private func openAccess(_ promise: Promise, walletOnly: Bool) {
      DispatchQueue.main.async {
        #if DEBUG
        guard #available(iOS 18.0, *), self.probe == nil, self.transfer == nil, !self.reconciling, let presenter = self.appContext?.utilities?.currentViewController(), let window = presenter.view.window else {
          promise.reject("UNAVAILABLE", "Native probe unavailable"); return
        }
        let probe = NativeProbe(presenter: presenter, window: window, walletOnly: walletOnly) { results in
          self.probe = nil
          switch results {
          case .success(.wallet(let address)): promise.resolve(["address": address, "accountIndex": 0, "chainId": 10143] as [String: Any])
          case .success(.proofs(let proofs)): promise.resolve(proofs.map { ["accountIndex": $0.accountIndex, "address": $0.address, "message": $0.message, "signature": $0.signatureHex] as [String: Any] })
          case .failure: promise.reject("PROBE_FAILED", "Native probe cancelled, unsupported or failed. Open existing after partial creation.")
          }
        }
        self.probe = probe
        probe.present()
        #else
        promise.reject("UNAVAILABLE", "Development probe only")
        #endif
      }
  }
  public func definition() -> ModuleDefinition {
    Name("GizuSigner")
    Function("getCapabilities") {
      #if DEBUG
      if #available(iOS 18.0, *) { return ["nativeTransfers": true, "fixedProbe": true] }
      #endif
      return ["nativeTransfers": false, "fixedProbe": false]
    }
    AsyncFunction("executeOperation") { (proposal: String, promise: Promise) in
      DispatchQueue.main.async {
        #if DEBUG
        guard #available(iOS 18.0, *), self.probe == nil, self.transfer == nil, !self.reconciling,
              let presenter = self.appContext?.utilities?.currentViewController(), presenter.view.window != nil else {
          promise.reject("UNAVAILABLE", "Native testnet operation unavailable"); return
        }
        do {
          try validateTransferProposal(proposal: proposal)
          let controller = NativeTransferController(proposal: proposal) { result in
            self.transfer = nil
            if let result { promise.resolve(result) }
            else { promise.reject("OPERATION_STOPPED", "Operation stopped. Refresh native status before retrying; submitted transfers cannot be undone.") }
          }
          self.transfer = controller; presenter.present(controller, animated: true)
        } catch { promise.reject("INVALID_OPERATION", "Native operation rejected") }
        #else
        promise.reject("UNAVAILABLE", "Development testnet operation only")
        #endif
      }
    }
    AsyncFunction("getOperationStatus") { (promise: Promise) in
      Task { @MainActor in
        guard self.probe == nil, self.transfer == nil, !self.reconciling else {
          promise.reject("BUSY", "Native operation in progress"); return
        }
        self.reconciling = true
        defer { self.reconciling = false }
        do { promise.resolve(try await NativeTransferJournal().reconcile(MonadRPC())) }
        catch { promise.reject("RPC_UNAVAILABLE", "Cannot reconcile native status. Do not repeat uncertain transfers.") }
      }
    }
    Function("cancelOperation") { DispatchQueue.main.async {
      if #available(iOS 18.0, *) { (self.transfer as? NativeTransferController)?.cancel() }
    } }
    Function("lock") { DispatchQueue.main.async {
      if #available(iOS 18.0, *) {
        (self.transfer as? NativeTransferController)?.cancel(); (self.probe as? NativeProbe)?.cancel()
      }
    } }
    AsyncFunction("openNativeProbe") { (promise: Promise) in self.openAccess(promise, walletOnly: false) }
    AsyncFunction("openWallet") { (promise: Promise) in self.openAccess(promise, walletOnly: true) }
    Function("cancelProbe") { DispatchQueue.main.async { if #available(iOS 18.0, *) { (self.probe as? NativeProbe)?.cancel() } } }
    OnDestroy { DispatchQueue.main.async { if #available(iOS 18.0, *) { (self.transfer as? NativeTransferController)?.cancel(); (self.probe as? NativeProbe)?.cancel() } } }
  }
}

private enum NativeAccessResult { case wallet(String); case proofs([ProbeProof]) }

@available(iOS 18.0, *)
private final class NativeProbe: NSObject, ASAuthorizationControllerDelegate, ASAuthorizationControllerPresentationContextProviding {
  private let presenter: UIViewController
  private let window: UIWindow
  private let walletOnly: Bool
  private var completion: ((Result<NativeAccessResult, Error>) -> Void)?
  private var controller: ASAuthorizationController?
  private var timeout: DispatchWorkItem?
  private var createdId: Data?
  private var alert: UIAlertController?
  private enum Failure: Error { case failed }
  init(presenter: UIViewController, window: UIWindow, walletOnly: Bool, completion: @escaping (Result<NativeAccessResult, Error>) -> Void) { self.presenter = presenter; self.window = window; self.walletOnly = walletOnly; self.completion = completion }
  private func random() throws -> Data {
    var data = Data(count: 32)
    let status = data.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }
    guard status == errSecSuccess else { throw Failure.failed }; return data
  }
  func present() {
    let alert = UIAlertController(title: walletOnly ? "Open Gizu testnet wallet" : "Native compatibility test", message: walletOnly ? "Create or open a passkey to view Account 0 on Monad testnet. No transaction or message will be signed. This opens a local wallet view, not a backend login. Creation may require two prompts." : "Approve 16 fixed test signatures across accounts 0–15 using an unfunded test passkey. No transactions, login or spending. Keys stay in native memory. Creation may show two prompts.", preferredStyle: .alert)
    alert.addAction(UIAlertAction(title: "Open existing", style: .default) { _ in self.request(create: false) })
    alert.addAction(UIAlertAction(title: walletOnly ? "Create passkey" : "Create test passkey", style: .default) { _ in self.request(create: true) })
    alert.addAction(UIAlertAction(title: "Cancel", style: .cancel) { _ in self.cancel() })
    self.alert = alert
    presenter.present(alert, animated: true)
    let work = DispatchWorkItem { self.cancel() }; timeout = work
    DispatchQueue.main.asyncAfter(deadline: .now() + 120, execute: work)
  }
  private func request(create: Bool) {
    guard completion != nil else { return }
    do {
      let provider = ASAuthorizationPlatformPublicKeyCredentialProvider(relyingPartyIdentifier: "gizu.io")
      let request: ASAuthorizationRequest
      if create {
        let registration = provider.createCredentialRegistrationRequest(challenge: try random(), name: walletOnly ? "gizu-testnet-wallet" : "gizu-native-test", userID: try random())
        registration.userVerificationPreference = .required
        registration.prf = .checkForSupport
        request = registration
      } else {
        let assertion = provider.createCredentialAssertionRequest(challenge: try random())
        assertion.userVerificationPreference = .required
        if let id = createdId { assertion.allowedCredentials = [ASAuthorizationPlatformPublicKeyCredentialDescriptor(credentialID: id)] }
        let hex = "896d46ac4ac191885c46137439db7bb52fb05cff3ecd34af7cdae0a1e0c00db9"
        let salt = Data(stride(from: 0, to: hex.count, by: 2).map { offset -> UInt8 in
          let start = hex.index(hex.startIndex, offsetBy: offset)
          return UInt8(hex[start..<hex.index(start, offsetBy: 2)], radix: 16)!
        })
        assertion.prf = .inputValues(.saltInput1(salt))
        request = assertion
      }
      controller = ASAuthorizationController(authorizationRequests: [request])
      controller?.delegate = self; controller?.presentationContextProvider = self
      controller?.performRequests()
    } catch { cancel() }
  }
  func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor { window }
  func authorizationController(controller: ASAuthorizationController, didCompleteWithAuthorization authorization: ASAuthorization) {
    guard completion != nil else { return }
    if let registration = authorization.credential as? ASAuthorizationPlatformPublicKeyCredentialRegistration { createdId = registration.credentialID; request(create: false); return }
    guard let credential = authorization.credential as? ASAuthorizationPlatformPublicKeyCredentialAssertion, let key = credential.prf?.first else { cancel(); return }
    // Never return credential/PRF to Expo. Swift/FFI copies remain a documented limitation.
    var bytes = key.withUnsafeBytes { Data($0) }
    defer { bytes.resetBytes(in: 0..<bytes.count) }
    do {
      guard UIApplication.shared.applicationState != .background else { throw Failure.failed }
      if walletOnly {
        finish(.success(.wallet(try deriveWalletAddress(prf: bytes))))
        return
      }
      let id = try random().map { String(format: "%02x", $0) }.joined()
      let result = try runNativeProbe(prf: bytes, operationId: id)
      finish(.success(.proofs(result)))
    } catch { cancel() }
  }
  func authorizationController(controller: ASAuthorizationController, didCompleteWithError error: Error) { cancel() }
  func cancel() { let pending = controller; finish(.failure(Failure.failed)); pending?.cancel(); alert?.dismiss(animated: false); alert = nil }
  private func finish(_ result: Result<NativeAccessResult, Error>) {
    timeout?.cancel(); timeout = nil
    let callback = completion; completion = nil; controller = nil; callback?(result)
  }
}
